pmkoalas/models/petrinet.py

```diff
@@ -0,0 +1,341 @@
+
+# 
+# Weighted Petri net with place and transition labels.
+# 
+# Allows explicit place or transition ids for simpler comparison, especially
+# during testing.
+#
+
+import operator
+import xml.etree.ElementTree as ET
+
+ENCODING='unicode'
+
+def steq(self,other):
+    if type(other) is type(self):
+        return self.__dict__ == other.__dict__
+    return False
+
+
+
+class LabelledPetriNet:
+    def __init__(self,places,transitions,arcs,label=None):
+        self._places = places
+        self._transitions = transitions
+        self._arcs = arcs
+        self._label = label
+
+    @property 
+    def places(self):
+        return self._places
+
+    @property
+    def transitions(self):
+        return self._transitions
+
+    @property
+    def arcs(self):
+        return self._arcs
+
+    @property
+    def label(self):
+        return self._label
+
+    def __eq__(self,other):
+        if isinstance(other,self.__class__):
+            return self._label  == other._label and \
+                   self._places == other._places and \
+                   self._transitions == other._transitions and \
+                   self._arcs   == self._arcs
+        return False
+
+    def reprcontents(self):
+        return f"places: {self._places} transitions: {self._transitions} arcs: {self._arcs}"
+
+    def __repr__(self):
+        return f"LabelledPetriNet:{self._label} " + self.reprcontents()
+
+    def __str__(self):
+        return "LabelledPetriNet(" + self._label + ")\n" \
+            +  " Places: " + str(self.places) + "\n" \
+            +  " Trans: " + str(self.transitions) + "\n" \
+            +  " Arcs: " + str(self.arcs)
+    
+
+class MutableLabelledPetriNet(LabelledPetriNet):
+    def __init__(self,label=None):
+        super(MutableLabelledPetriNet,self).__init__(set(),set(),set(),label)
+
+    def addPlace(self,place):
+        self._places.add(place)
+
+    def addTransition(self,tran):
+        self._transitions.add(tran)
+
+    def addArc(self,arc):
+        self._arcs.add(arc)
+
+    def addArcForNodes(self,fromNode,toNode):
+        self.addArc(Arc(fromNode,toNode))
+
+    def __repr__(self):
+        return f"MutableLabelledPetriNet:{self._label} " + self.reprcontents()
+
+    def __eq__(self,other):
+        if isinstance(other,LabelledPetriNet):
+            return self._label  == other._label and \
+                   self._places == other._places and \
+                   self._transitions == other._transitions and \
+                   self._arcs   == self._arcs
+        return False
+
+def verbosecmp(obj1,obj2):
```

Comment:
I can see that you have used this function a lot in your tests but should this be in the library or in the testing module? Not really sure what this does

Reply:
It gives a verbose description of differences between two of these objects. It is useful in a functional testing context as well as unit tests.